### PR TITLE
[AST] NFC: Add reentrancy check to associated type anchor walk

### PR DESCRIPTION
**Pull request by davezarzycki**
This makes compiler crasher 28870 crash faster and reliably instead of recursing off the end of the stack.

This reverts: 614cb48b63f29325e5b7aee267602c2c27f1a3c6

### CODE REVIEW

**Comment by davezarzycki** [comment]
@swift-ci please smoke test

**Comment by davezarzycki** [comment]
Hi @DougGregor and @slavapestov  – I just tried adding an assert to `GenericSignatureBuilder::addInheritedRequirements()` to ensure that a protocol's inheritance is known acyclic, but that blows up because the inheritance of a protocol can be used before the protocol has been type checked.

I'd like to keep the assert or something like it. Can we do protocol type checking earlier? Or does it need to be phased? In other words, validate that inheritances are acyclic "early" somehow?

**Comment by DougGregor** [comment]
I'd prefer if we could fix this crash by ensuring that the request-evaluator is detecting and coping with the cycle, rather than updating that early cycle checking (which we can't rely on in general). Maybe it's because of the dodgy hack I implemented in https://github.com/apple/swift/blob/master/lib/AST/Decl.cpp#L3229-L3236 for performance reasons, or maybe it's something different, but doing cycle checking outside of the request-evaluator is a game of whack-a-mole that I'd hoped we could avoid doing more of.

**Comment by davezarzycki** [comment]
Maybe the "dodgy hack" is the problem then? Compiler crash 28870 recurses off the end of the stack via two cyclicly recursive protocols that try to call `getOverriddenDecls()` on each other.

File: lib/Sema/TypeCheckDecl.cpp [hunk 1]
```diff
@@ -3151,25 +3155,26 @@ class DeclChecker : public DeclVisitor<DeclChecker> {
 
     checkUnsupportedNestedType(PD);
 
+    // Check for circular inheritance within the protocol.
+    SmallVector<ProtocolDecl *, 8> path;
+    path.push_back(PD);
+    checkCircularity(TC, PD, diag::circular_protocol_def,
+                     DescriptiveDeclKind::Protocol, path);
+    // Skip full validation if circularity exists.
+    if (PD->isInvalid())
```
Comment (slavapestov) [inline lib/Sema/TypeCheckDecl.cpp]: Someone else can still call validateDecl() on the protocol, though, for example if a name lookup references a protocol in another file. This is not really the right fix unfortunately.

**Comment by davezarzycki** [comment]
@DougGregor and @slavapestov – I've reworked `AssociatedTypeDecl::getAssociatedTypeAnchor()` to not crash when circular (and erroneous) protocols exist.

@swift-ci please smoke test

**Comment by davezarzycki** [comment]
Ping? I'd like to merge this because compiler crasher 28870 times out semi-regularly these days on my Linux box, which impacts my automated testing. Thanks.

**Comment by slavapestov** [comment]
Sorry for not replying sooner. I think this function would be clearer if it was non-recursive. Other than that it looks like a good fix for now.

**Pull request merged**